### ISSUE TITLE
Lets make it practical

### ISSUE THREAD

**Issue by simsim314** [Zeta36/chess-alpha-zero#32]
I try to use your code to train a model, but there are several "stop production" issues: 

1. You've mentioned already two planes is not enough, but is it stop production? I think it's critical. 
2. I saw you're using two layer CNN model, while DeepMind usually use very deep network. How complex a model should be, to be capable to learn chess? 
3. Generating self play takes too long. I have good GPU (GTX 760) and it takes me a minute per game. While we need tens of millions of games, generating 2K games takes 24 hours (it will take about 30 years to only generate all the data). 
4. Even supervised learning has limitation. The biggest limitation is loading all the games into memory, before GPU optimization. I've 8GB RAM this makes me limited to 3K games, how about loading 1K games running about 5 epochs, and then load new games. This will allow to train on tens of thousand of games.

**Comment by Akababa** [comment]
Hi @simsim314, thanks for sharing your thoughts.

This project is still under active construction, just yesterday I wrote a [much faster MCTS](https://github.com/Akababa/chess-alpha-zero/issues/3), and I'll optimize and test it more today. As @Zeta36 said, you're always welcome to contribute and ask questions here if you need any help.

**Comment by benediamond** [comment]
@simsim314, In a fork, I have implemented:
1) The DeepMind-style 119 planes of input (see [here](https://github.com/benediamond/chess-alpha-zero/blob/af7caee80929193522afb30fbd66b4ca56935b14/src/chess_zero/env/chess_env.py#L177)).
2) The DeepMind-style NN architecture, with 19 residual layers (see [here](https://github.com/benediamond/chess-alpha-zero/blob/af7caee80929193522afb30fbd66b4ca56935b14/src/chess_zero/agent/model_chess.py#L38)).
3) ...see Akababa's comment...!
4) Loading only the _n_ most recent play data files are loaded into memory during optimization, on a "rolling" basis, to ease memory consumption (see [here](https://github.com/benediamond/chess-alpha-zero/blob/af7caee80929193522afb30fbd66b4ca56935b14/src/chess_zero/worker/optimize.py#L120)).

Unfortunately, in this setup I have _failed to achieve convergence_, even during supervised learning. Please feel free to help investigate the reason why.

**Comment by Zeta36** [comment]
@benediamond, could you try a fast check?

Why don't you add (as @simsim314 said) to the model some more CNN layers like this:

```
        x = Conv2D(filters=mc.cnn_filter_num, kernel_size=mc.cnn_filter_size, padding="same",
                   data_format="channels_first", kernel_regularizer=l2(mc.l2_reg))(x)
        x = BatchNormalization(axis=1)(x)
        x = Activation("relu")(x)
```

before applying the residual blocks?

Maybe the lack of convergence in your NN with so many feed planes is due to the current limitation to two Conv2D layers in our model.

**Comment by benediamond** [comment]
@Zeta36 I'm not sure I understand. As it stands, following DeepMind, we _already_ have a residual tower consisting of

1) A convolutional layer containing two filters
2) 19 residual blocks, _each of which_ contains two convolutional layer of 256 filters each.

There are then further convolutions in the policy and value heads.

Let me know if you still think something should be changed.

**Comment by Zeta36** [comment]
You are right, @benediamond. We have already two convolutional layer for each residual block. My mistake. I don't really know why nor your model neither mine are able to converge when we introduce new feed planes :(. 

I could not even make to converge a 14 planes feeding with your one-hot encoder pieces development (??). I wonder if we could at least converge a model with some more lineal planes (without one-hot encoder) like current player color, number of movement, etc. but leaving the pieces planes as ord() integer values.

**Comment by Akababa** [comment]
@benediamond Have you tried playing with your model yourself to see if it's qualitatively "getting better"? I'm worried about us falling into the trap of mixing validation and training data. Also I don't know if the concept of losses converging to 0 is a sound one, because a) you have regularization and b) it can't be lower than the "shannon entropy" of the training data, if that makes any sense.

**Comment by simsim314** [comment]
OK I see everything's there, except that it doesn't work well. I think from practical standpoint, we should reach a place where alpha-zero is not giving away it's queen or pieces for free.

How about using some engine, to train alpha-zero on its blunders in certain positions instead of games, this will reduce the training noise significantly. Once we reach a point where it starts to play reasonably well, we can use self-play to improve.

**Comment by benediamond** [comment]
@simsim314 see the comments on [this thread](https://github.com/Akababa/chess-alpha-zero/issues/3). I'm working on a new version that addresses the "policy flipping" issue; I think Akababa might already have one.

**Comment by Akababa** [comment]
@simsim314 Good ideas, I'm currently adjudicating games based on material but as you say it's probably faster to train the naked network on stockfish outputs or something. If you're doing any self play or evaluation though I have a [multithreaded MCTS implementation](https://github.com/Akababa/chess-alpha-zero/blob/master/src/chess_zero/agent/player_chess.py) which is much faster.

I started a wiki page on [supervised methods](https://github.com/Akababa/chess-alpha-zero/wiki/Ways-to-cheat-in-training) so we can organize our thoughts

**Comment by simsim314** [comment]
I think we should be realistic about our access to good hardware. Google had used 5000 TPUs to generate self play, I think we can safely assume we will not get something like this soon. So we should focus on using existing games, and even there making the best out of them - because running on 40 million games with single GPU will currently take years. So we probably need to analyze blunders in positions and teach our network to avoid making mistakes. 

Another point is that MCTS AlphaZero is using, runs 80K variation in second, using 4 TPUs. This is equivalent to 720 TFLOPs, or about 100-200 strong GPUs. On my GPU it runs ~800 variations in second and the question is whether it's possible to use such low count of simulations, to even get something that plays well, aiming to play as good as some engines (above 2500).

**Comment by Akababa** [comment]
I think ideally we want to follow a path similar to leela zero; find an architecture that shows good potential (which is arguably applying human heuristics lol), write a robust implementation, and start a distributed effort. It's really quite straightforward, but at the moment we're still trying to fix bugs, validate the models and wait for more contributors... tomorrow I'll start looking into [C++ implementation](https://github.com/benediamond/chess-alpha-zero/pull/16#issuecomment-352319494) of this project, you're welcome to join in if you want to expedite the process.

**Comment by lucasart** [comment]
@Zeta36: Regarding speed, I expect the bottleneck is in gameplay, written in Python. I am happy to help you with a minimal C implementation for that. Let me know if you're interested. I think your code is beautiful, and I rewriting all of it in C++ is a bad idea, but just a C portion for speed critical gameplay seems appropriate.

**Comment by Zeta36** [comment]
Yes!! of course your collaboration will be welcome. Please checkout and as soon as you get stable results ask for a pull request :).
